platform/engineering/design-docs/2021-01-13-web-component-library.md

```diff
@@ -0,0 +1,139 @@
+
+Until we reach that point, the WC branch will be developed in the `component-library` repo alongside the main branch containing the React components. The WC branch will not be published to npm, and the `vets-website` dependency will instead use a Git url with a tag.
+
+This WC-specific branch will use [Stencil](https://stenciljs.com/) as a tool to help build the Components, and it will also have [Storybook](https://storybook.js.org/) integration. Updates to this branch will follow the typical development flow using PRs. Anytime a new WC is developed or a significant change is made, a new release will be tagged.
+
+One of the additional benefits of Web Components is how they can be styled. Each Component can have its own styles scoped to its [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) (this means that if a Component renders a `<div>`, the CSS can broadly style the `div` node without worrying about it affecting all the other `<div>`s on the page). As a result, each component will have it's own CSS file which also makes use of global [CSS variables](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties).
+
+Once in `vets-website`, the setup code for the WC library will be in its own entry file that is loaded on each page. In addition, this script will be one of the first loaded in order to ensure that the Components will be defined in the browser before they are rendered on the page.
+
+### Code Location
+
+The code will live in the `component-library` repo on a dedicated orphan branch named `web-components`. This code will be used in `vets-website` via a dependency in `package.json`
+
+### Testing Plan
+
+Each Component will have:
+
+- Unit and E2E tests written using Jest
+- One or more stories in Storybook to validate appearance/behavior
+
+As of mid-January 2021, Stencil provides an experimental screenshot visual diff testing functionality. If this becomes stable, we will consider adopting it.
+
+### Logging
+
+N/A
+
+### Debugging
+
+Storybook will be a useful tool for debugging and development. Stencil also comes with its own development server which can be used.
+
+There will be no special provisions for debugging provided for consumers of the WC library.
+
+### Caveats
+
+To be determined.
+
+### Security Concerns
+
+N/A
+
+### Privacy Concerns
+
+N/A
+
+### Open Questions and Risks
+
+Accessibility and performance are two things that should receive lots of attention.
+
+Do we want to remove the explicit dependency on [USWDS](https://designsystem.digital.gov/)? Rather than explicitly relying on `uswds`, the Component library could be an abstraction for `vets-website` that implements the appropriate styles from `uswds`. A Component library complete with global styling defaults might be able to accomplish everything that `uswds` is providing.
```

Comment:
Two things come to mind:
1. Because we can independently style web components, is it feasible to adopt USWDS 2.0 styles in just our web components?
1. I believe the design system's component library should be an implementation of the visual design language that the design system defines. It should be opinionated about its styles and provide overrides only as needed. (To maintain consistency across all of VA.gov, those needs should be deliberate—more than just, "I don't like it.")

Reply:
1. The short answer is "probably". The longer answer would involve looking at the differences between USWDS 1.0 and 2.0 and seeing where they differ. However, we can make the CSS for each component be whatever we want. This could mean copy-pasting USWDS stuff (not ideal), or adding it as a dependency and `@import`ing the appropriate CSS files into each component and then making any additional modifications that we want.

With that being said, I think that adopting USWDS 2.0 as part of this design document would fall under non-goals:

> A non-goal is creating an entirely new visual experience as part of the component library.

2. I agree.

Reply:
My opinion: Let's say we're going to use USWDS 2.0 for our WC styling and if we run into any problems with that, we can reassess later. The design document is meant to be updated if we find the implementation needs to change.